### ISSUE TITLE
Core dump in bleadperl from perl -e '$! = 9999; $a = $!;'

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10178]
Migrated from [rt.perl.org#72850](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72850) (status was 'resolved')

Searchable as RT72850$

**Comment by p5pRT** [comment]
### From @timbunce



perl \-e '$\! = 9999; $a = $\!;'
Bus error \(core dumped\)

\#0  0x0012e008 in Perl\_magic\_get \(my\_perl=0x1000000\, sv=0x10489f0\, mg=0x847930\) at mg\.c&#8203;:1055
1055        SvRTRIM\(sv\);
\(gdb\) bt
\#0  0x0012e008 in Perl\_magic\_get \(my\_perl=0x1000000\, sv=0x10489f0\, mg=0x847930\) at mg\.c&#8203;:1055
\#1  0x00126d90 in Perl\_mg\_get \(my\_perl=0x1000000\, sv=0x10489f0\) at mg\.c&#8203;:225
\#2  0x00194f36 in Perl\_sv\_2nv \(my\_perl=0x1000000\, sv=0x10489f0\) at sv\.c&#8203;:2463
\#3  0x00166a66 in Perl\_pp\_add \(my\_perl=0x1000000\) at pp\_hot\.c&#8203;:639
\#4  0x00110ce1 in Perl\_runops\_debug \(my\_perl=0x1000000\) at dump\.c&#8203;:2049
\#5  0x00038135 in S\_run\_body \(my\_perl=0x1000000\, oldscope=1\) at perl\.c&#8203;:2308
\#6  0x000371ce in perl\_run \(my\_perl=0x1000000\) at perl\.c&#8203;:2233
\#7  0x000021d5 in main \(argc=3\, argv=0xbfffeedc\, env=0xbfffeeec\) at perlmain\.c&#8203;:117

\(gdb\) print \*sv
$1 = \{
&nbsp;  sv\_any = 0x1024c74\, 
&nbsp;  sv\_refcnt = 2\, 
&nbsp;  sv\_flags = 17415\, 
&nbsp;  sv\_u = \{
&nbsp;    svu\_pv = 0x0\, 
&nbsp;    svu\_iv = 0\, 
&nbsp;    svu\_uv = 0\, 
&nbsp;    svu\_rv = 0x0\, 
&nbsp;    svu\_array = 0x0\, 
&nbsp;    svu\_hash = 0x0\, 
&nbsp;    svu\_gp = 0x0
&nbsp;  \}
\}

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=critical

Site configuration information for perl 5.11.4:

Configured by timbo at Wed Feb 10 15:52:10 GMT 2010.

Summary of my perl5 (revision 5 version 11 subversion 4) configuration:
  Commit id: 2630fd9e8e31d2fd409e2e8ec16dc85d230a3eb3
  Platform:
    osname=darwin, osvers=10.2.0, archname=darwin-thread-multi-2level
    uname='darwin timac.local 10.2.0 darwin kernel version 10.2.0: tue nov 3 10:37:10 pst 2009; root:xnu-1486.2.11~1release_i386 i386 '
    config_args='-des -Doptimize=-g -DEBUGGING=both -Dusethreads -Dusemultiplicity -Dusedevel -Uversiononly -Dprefix=~/blead'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-common -DPERL_DARWIN -no-cpp-precomp -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -I/opt/local/include',
    optimize='-g',
    cppflags='-no-cpp-precomp -fno-common -DPERL_DARWIN -no-cpp-precomp -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -I/opt/local/include'
    ccversion='', gccversion='4.2.1 (Apple Inc. build 5646)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector -L/usr/local/lib -L/opt/local/lib'
    libpth=/usr/local/lib /opt/local/lib /usr/lib
    libs=-lgdbm -ldbm -ldl -lm -lutil -lc
    perllibs=-ldl -lm -lutil -lc
    libc=/usr/lib/libc.dylib, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/usr/local/lib -L/opt/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.11.4:
    /Users/timbo/blead/lib/perl5/site_perl/5.11.4/darwin-thread-multi-2level
    /Users/timbo/blead/lib/perl5/site_perl/5.11.4
    /Users/timbo/blead/lib/perl5/5.11.4/darwin-thread-multi-2level
    /Users/timbo/blead/lib/perl5/5.11.4
    /Users/timbo/blead/lib/perl5/site_perl/5.10.1
    /Users/timbo/blead/lib/perl5/site_perl/5.10.0
    /Users/timbo/blead/lib/perl5/site_perl
    .


Environment for perl 5.11.4:
    HOME=/Users/timbo
    LANG=en_IE.UTF-8
    LANGUAGE (unset)
    LC_ALL=en_IE.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/Users/timbo/blead/bin:/usr/local/perl5-blead-deb-thr/bin:/usr/local/openjdk6-b16-24_apr_2009-r1/bin:/opt/local/lib/postgresql83/bin:/Users/timbo/bin:/usr/local/perl58-i/bin:/usr/local/mysql/bin:/usr/local/bin:/opt/local/bin:/opt/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/usr/X11/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

On 15 February 2010 15&#8203;:06\, Tim Bunce \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> perl \-e '$\! = 9999; $a = $\!;'
> Bus error \(core dumped\)

I do not reproduce this ; what does strerror return for errno 9999 on
your system?
On my Linux it's&#8203;:
$ \./perl \-le 'print $\!=9999'
Unknown error 9999
and gdb with a breakpoint at mg\.c&#8203;:1055 confirms this\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, Feb 15\, 2010 at 06&#8203;:06&#8203;:50AM \-0800\, Tim Bunce wrote&#8203;:

> perl \-e '$\! = 9999; $a = $\!;'
> Bus error \(core dumped\)
> 
> \#0  0x0012e008 in Perl\_magic\_get \(my\_perl=0x1000000\, sv=0x10489f0\, mg=0x847930\) at mg\.c&#8203;:1055
> 1055        SvRTRIM\(sv\);
> \(gdb\) bt
> \#0  0x0012e008 in Perl\_magic\_get \(my\_perl=0x1000000\, sv=0x10489f0\, mg=0x847930\) at mg\.c&#8203;:1055
> \#1  0x00126d90 in Perl\_mg\_get \(my\_perl=0x1000000\, sv=0x10489f0\) at mg\.c&#8203;:225
> \#2  0x00194f36 in Perl\_sv\_2nv \(my\_perl=0x1000000\, sv=0x10489f0\) at sv\.c&#8203;:2463
> \#3  0x00166a66 in Perl\_pp\_add \(my\_perl=0x1000000\) at pp\_hot\.c&#8203;:639
> \#4  0x00110ce1 in Perl\_runops\_debug \(my\_perl=0x1000000\) at dump\.c&#8203;:2049
> \#5  0x00038135 in S\_run\_body \(my\_perl=0x1000000\, oldscope=1\) at perl\.c&#8203;:2308
> \#6  0x000371ce in perl\_run \(my\_perl=0x1000000\) at perl\.c&#8203;:2233
> \#7  0x000021d5 in main \(argc=3\, argv=0xbfffeedc\, env=0xbfffeeec\) at perlmain\.c&#8203;:117
> 
> \(gdb\) print \*sv
> $1 = \{
>   sv\_any = 0x1024c74\, 
>   sv\_refcnt = 2\, 
>   sv\_flags = 17415\, 

0x4407

PVMG\, POK\, pPOK

>   sv\_u = \{
>     svu\_pv = 0x0\, 

NULL\. That's naughty\. That SV is not well formed\. I wonder\, how did that come
to be?

Try the appended patch\. Tested on Linux\, but not OS X

> Summary of my perl5 \(revision 5 version 11 subversion 4\) configuration&#8203;:
>   Commit id&#8203;: 2630fd9e8e31d2fd409e2e8ec16dc85d230a3eb3
>   Platform&#8203;:
>     osname=darwin\, osvers=10\.2\.0\, archname=darwin\-thread\-multi\-2level

Nicholas Clark

<details><summary>Inline Patch</summary>

```diff
diff --git a/mg.c b/mg.c
index cc01547..06c899e 100644
--- a/mg.c
+++ b/mg.c
@@ -1048,7 +1048,8 @@ Perl_magic_get(pTHX_ SV *sv, MAGIC *mg)
 	else
 #endif
 	sv_setpv(sv, errno ? Strerror(errno) : "");
-	SvPOK_on(sv);	/* may have got removed during taint processing */
+	if (SvPOKp(sv))
+	    SvPOK_on(sv);    /* may have got removed during taint processing */
 	RESTORE_ERRNO;
 	}
 
```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

On Tue\, Feb 16\, 2010 at 3&#8203;:59 AM\, Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:
> On Mon\, Feb 15\, 2010 at 06&#8203;:06&#8203;:50AM \-0800\, Tim Bunce wrote&#8203;:
>
>> perl \-e '$\! = 9999; $a = $\!;'
>> Bus error \(core dumped\)
>>
>> \#0  0x0012e008 in Perl\_magic\_get \(my\_perl=0x1000000\, sv=0x10489f0\, mg=0x847930\) at mg\.c&#8203;:1055
>> 1055        SvRTRIM\(sv\);
>> \(gdb\) bt
>> \#0  0x0012e008 in Perl\_magic\_get \(my\_perl=0x1000000\, sv=0x10489f0\, mg=0x847930\) at mg\.c&#8203;:1055
>> \#1  0x00126d90 in Perl\_mg\_get \(my\_perl=0x1000000\, sv=0x10489f0\) at mg\.c&#8203;:225
>> \#2  0x00194f36 in Perl\_sv\_2nv \(my\_perl=0x1000000\, sv=0x10489f0\) at sv\.c&#8203;:2463
>> \#3  0x00166a66 in Perl\_pp\_add \(my\_perl=0x1000000\) at pp\_hot\.c&#8203;:639
>> \#4  0x00110ce1 in Perl\_runops\_debug \(my\_perl=0x1000000\) at dump\.c&#8203;:2049
>> \#5  0x00038135 in S\_run\_body \(my\_perl=0x1000000\, oldscope=1\) at perl\.c&#8203;:2308
>> \#6  0x000371ce in perl\_run \(my\_perl=0x1000000\) at perl\.c&#8203;:2233
>> \#7  0x000021d5 in main \(argc=3\, argv=0xbfffeedc\, env=0xbfffeeec\) at perlmain\.c&#8203;:117
>>
>> \(gdb\) print \*sv
>> $1 = \{
>>   sv\_any = 0x1024c74\,
>>   sv\_refcnt = 2\,
>>   sv\_flags = 17415\,
>
> 0x4407
>
> PVMG\, POK\, pPOK
>
>>   sv\_u = \{
>>     svu\_pv = 0x0\,
>
> NULL\. That's naughty\. That SV is not well formed\. I wonder\, how did that come
> to be?
>
> Try the appended patch\. Tested on Linux\, but not OS X
>
>> Summary of my perl5 \(revision 5 version 11 subversion 4\) configuration&#8203;:
>>   Commit id&#8203;: 2630fd9e8e31d2fd409e2e8ec16dc85d230a3eb3
>>   Platform&#8203;:
>>     osname=darwin\, osvers=10\.2\.0\, archname=darwin\-thread\-multi\-2level
>
> Nicholas Clark
>
> diff \-\-git a/mg\.c b/mg\.c
> index cc01547\.\.06c899e 100644
> \-\-\- a/mg\.c
> \+\+\+ b/mg\.c
> @&#8203;@&#8203; \-1048\,7 \+1048\,8 @&#8203;@&#8203; Perl\_magic\_get\(pTHX\_ SV \*sv\, MAGIC \*mg\)
>        else
>  \#endif
>        sv\_setpv\(sv\, errno ? Strerror\(errno\) : ""\);
> \-       SvPOK\_on\(sv\);   /\* may have got removed during taint processing \*/
> \+       if \(SvPOKp\(sv\)\)
> \+           SvPOK\_on\(sv\);    /\* may have got removed during taint processing \*/
>        RESTORE\_ERRNO;
>        \}
>

That fixes it for me on OS X\.

Steve

**Comment by p5pRT** [comment]
### From @smpeters

On Tue\, Feb 16\, 2010 at 9&#8203;:12 AM\, Steve Peters \<steve@&#8203;fisharerojo\.org> wrote&#8203;:
> On Tue\, Feb 16\, 2010 at 3&#8203;:59 AM\, Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:
>> On Mon\, Feb 15\, 2010 at 06&#8203;:06&#8203;:50AM \-0800\, Tim Bunce wrote&#8203;:
>>
>>> perl \-e '$\! = 9999; $a = $\!;'
>>> Bus error \(core dumped\)
>>>
>>> \#0  0x0012e008 in Perl\_magic\_get \(my\_perl=0x1000000\, sv=0x10489f0\, mg=0x847930\) at mg\.c&#8203;:1055
>>> 1055        SvRTRIM\(sv\);
>>> \(gdb\) bt
>>> \#0  0x0012e008 in Perl\_magic\_get \(my\_perl=0x1000000\, sv=0x10489f0\, mg=0x847930\) at mg\.c&#8203;:1055
>>> \#1  0x00126d90 in Perl\_mg\_get \(my\_perl=0x1000000\, sv=0x10489f0\) at mg\.c&#8203;:225
>>> \#2  0x00194f36 in Perl\_sv\_2nv \(my\_perl=0x1000000\, sv=0x10489f0\) at sv\.c&#8203;:2463
>>> \#3  0x00166a66 in Perl\_pp\_add \(my\_perl=0x1000000\) at pp\_hot\.c&#8203;:639
>>> \#4  0x00110ce1 in Perl\_runops\_debug \(my\_perl=0x1000000\) at dump\.c&#8203;:2049
>>> \#5  0x00038135 in S\_run\_body \(my\_perl=0x1000000\, oldscope=1\) at perl\.c&#8203;:2308
>>> \#6  0x000371ce in perl\_run \(my\_perl=0x1000000\) at perl\.c&#8203;:2233
>>> \#7  0x000021d5 in main \(argc=3\, argv=0xbfffeedc\, env=0xbfffeeec\) at perlmain\.c&#8203;:117
>>>
>>> \(gdb\) print \*sv
>>> $1 = \{
>>>   sv\_any = 0x1024c74\,
>>>   sv\_refcnt = 2\,
>>>   sv\_flags = 17415\,
>>
>> 0x4407
>>
>> PVMG\, POK\, pPOK
>>
>>>   sv\_u = \{
>>>     svu\_pv = 0x0\,
>>
>> NULL\. That's naughty\. That SV is not well formed\. I wonder\, how did that come
>> to be?
>>
>> Try the appended patch\. Tested on Linux\, but not OS X
>>
>>> Summary of my perl5 \(revision 5 version 11 subversion 4\) configuration&#8203;:
>>>   Commit id&#8203;: 2630fd9e8e31d2fd409e2e8ec16dc85d230a3eb3
>>>   Platform&#8203;:
>>>     osname=darwin\, osvers=10\.2\.0\, archname=darwin\-thread\-multi\-2level
>>
>> Nicholas Clark
>>
>> diff \-\-git a/mg\.c b/mg\.c
>> index cc01547\.\.06c899e 100644
>> \-\-\- a/mg\.c
>> \+\+\+ b/mg\.c
>> @&#8203;@&#8203; \-1048\,7 \+1048\,8 @&#8203;@&#8203; Perl\_magic\_get\(pTHX\_ SV \*sv\, MAGIC \*mg\)
>>        else
>>  \#endif
>>        sv\_setpv\(sv\, errno ? Strerror\(errno\) : ""\);
>> \-       SvPOK\_on\(sv\);   /\* may have got removed during taint processing \*/
>> \+       if \(SvPOKp\(sv\)\)
>> \+           SvPOK\_on\(sv\);    /\* may have got removed during taint processing \*/
>>        RESTORE\_ERRNO;
>>        \}
>>
>
> That fixes it for me on OS X\.
>
> Steve
>

Here's a test case for the fix as well\.

Steve

\-\-\- a/t/op/magic\.t
\+\+\+ b/t/op/magic\.t
@&#8203;@&#8203; \-12\,7 \+12\,7 @&#8203;@&#8203; BEGIN \{
 use warnings;
 use Config;

\-plan \(tests => 79\);
\+plan \(tests => 80\);

 $Is\_MSWin32  = $^O eq 'MSWin32';
 $Is\_NetWare  = $^O eq 'NetWare';
@&#8203;@&#8203; \-491\,3 \+491\,9 @&#8203;@&#8203; foreach my $sig \(qw\(\_\_DIE\_\_ \_BOGUS\_HOOK KILL THIRSTY\)\) \{
&nbsp;     is $SIG\{$sig\}\, undef\, "$sig is not present";
&nbsp;     is delete $SIG\{$sig\}\, undef\, "delete of $sig returns undef";
 \}
\+
\+\{
\+    $\! = 9999;
\+    is int $\!\, 9999\, q\{\[perl \#72850\] Core dump in bleadperl from perl
\-e '$\! = 9999; $a = $\!;'\};
\+
\+\}

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Feb 16\, 2010 at 09&#8203;:25&#8203;:21AM \-0600\, Steve Peters wrote&#8203;:
> On Tue\, Feb 16\, 2010 at 9&#8203;:12 AM\, Steve Peters \<steve@&#8203;fisharerojo\.org> wrote&#8203;:
> > On Tue\, Feb 16\, 2010 at 3&#8203;:59 AM\, Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:

> >> Try the appended patch\. Tested on Linux\, but not OS X
> >>
> >>> Summary of my perl5 \(revision 5 version 11 subversion 4\) configuration&#8203;:
> >>>   Commit id&#8203;: 2630fd9e8e31d2fd409e2e8ec16dc85d230a3eb3
> >>>   Platform&#8203;:
> >>>     osname=darwin\, osvers=10\.2\.0\, archname=darwin\-thread\-multi\-2level
> >>
> >> Nicholas Clark
> >>
> >> diff \-\-git a/mg\.c b/mg\.c
> >> index cc01547\.\.06c899e 100644
> >> \-\-\- a/mg\.c
> >> \+\+\+ b/mg\.c
> >> @&#8203;@&#8203; \-1048\,7 \+1048\,8 @&#8203;@&#8203; Perl\_magic\_get\(pTHX\_ SV \*sv\, MAGIC \*mg\)
> >>        else
> >>  \#endif
> >>        sv\_setpv\(sv\, errno ? Strerror\(errno\) : ""\);
> >> \-       SvPOK\_on\(sv\);   /\* may have got removed during taint processing \*/
> >> \+       if \(SvPOKp\(sv\)\)
> >> \+           SvPOK\_on\(sv\);    /\* may have got removed during taint processing \*/
> >>        RESTORE\_ERRNO;
> >>        \}
> >>
> >
> > That fixes it for me on OS X\.
> >
> > Steve
> >
> 
> Here's a test case for the fix as well\.

Thanks\, applied as part of be1cf43c8dab9dd236839206d53611f7e7d2d856

Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

Fixed in be1cf43c8dab9dd236839206d53611f7e7d2d856

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'open' to 'resolved'
